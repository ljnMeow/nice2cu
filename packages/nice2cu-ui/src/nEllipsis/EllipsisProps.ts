import { ExtractPropTypes, PropType } from 'vue';

function directionValidator(direction: string): boolean {
	return ['start', 'center', 'end'].includes(direction);
}

export const EllipsisProps = {
	content: {
		type: String,
		default: '',
	},
	lineHeight: {
		type: String,
		default: '20px',
	},
	direction: {
		type: String as PropType<'start' | 'center' | 'end'>,
		validator: directionValidator,
		default: 'end',
	},
	rows: {
		type: Number,
		default: 1,
	},
	showExpand: {
		type: Boolean,
		default: false,
	},
	expandText: {
		type: String,
		default: '展开',
	},
	collapseText: {
		type: String,
		default: '收起',
	},
};

export type EllipsisPropsType = ExtractPropTypes<typeof EllipsisProps>;
