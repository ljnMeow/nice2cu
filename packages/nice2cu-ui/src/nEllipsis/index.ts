import _ellipsis from './Ellipsis.vue';
import { withInstall } from '../../utils/withInstall';

const nEllipsis = withInstall(_ellipsis);

export default nEllipsis;

export { nEllipsis };

declare module 'vue' {
	export interface GlobalComponents {
		nEllipsis: typeof nEllipsis;
	}
}
