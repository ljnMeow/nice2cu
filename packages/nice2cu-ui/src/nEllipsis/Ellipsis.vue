<template>
	<div ref="ellipsis" :class="[bem.b()]">
		<div v-if="!isCalculate" ref="calculate" :class="[bem.m('calculate')]">{{ calculateText }}</div>
		<template v-if="!exceeded">
			{{ content }}
		</template>
		<template v-if="exceeded && !expanded && isCalculate">
			{{ contentObj && contentObj.endText }}<span v-if="showExpand" :class="bem.e('expandText')" @click.stop="handlerClick(1)">{{ expandText }}</span
			>{{ contentObj && contentObj.startText }}
		</template>
		<template v-if="exceeded && expanded">
			{{ content }}<span v-if="showExpand" :class="bem.e('expandText')" @click.stop="handlerClick(2)">{{ collapseText }}</span>
		</template>
	</div>
</template>

<script lang="ts">
import { defineComponent, ref, onMounted, reactive, watch, toRefs } from 'vue';
import { createNamespace } from '../../utils/create';
import { convertToNumber, waitingScreenRedrawn } from '../../utils/tools';
import { EllipsisProps, EllipsisPropsType } from './EllipsisProps';
import './style/ellipsis.less';

export default defineComponent({
	name: 'NEllipsis',
	props: EllipsisProps,
	setup(props: EllipsisPropsType) {
		const bem = createNamespace('ellipsis');

		const ellipsis = ref<HTMLElement | null>(null);
		const calculate = ref<HTMLElement | null>(null);
		const isCalculate = ref<boolean>(false);
		const calculateText = ref(props.content);
		const state = reactive({
			exceeded: false,
			expanded: false,
			contentObj: { startText: '', endText: '' } as { startText?: string; endText?: string },
		});
		const dot = '...';
		let maxHeight = 0;

		const calcEllipse = async () => {
			if (!ellipsis.value || !calculate.value) return;
			const originStyle = window.getComputedStyle(calculate.value);
			const lineHeight = convertToNumber(originStyle.lineHeight === 'normal' ? props.lineHeight : originStyle.lineHeight);

			maxHeight = Math.floor(
				lineHeight * (Number(props.rows) + 0.5) + convertToNumber(originStyle.paddingTop) + convertToNumber(originStyle.paddingBottom)
			);

			if (calculate.value.offsetHeight <= maxHeight) {
				state.exceeded = false;
				isCalculate.value = true;
			} else {
				state.exceeded = true;
				const end = props.content.length;
				const middle = Math.floor((0 + end) / 2);
				state.contentObj = props.direction === 'center' ? await cuttingMiddleText([0, middle], [middle, end]) : await cuttingText(0, end);
				isCalculate.value = true;
			}
		};

		const cuttingMiddleText = async (leftPart: [number, number], rightPart: [number, number]): Promise<{ startText?: string; endText?: string }> => {
			const actionText = props.showExpand ? (state.expanded ? props.collapseText : props.expandText) : '';
			const end = props.content.length;

			if (leftPart[1] - leftPart[0] <= 1 && rightPart[1] - rightPart[0] <= 1) {
				return {
					endText: props.content.slice(0, leftPart[0]) + dot,
					startText: dot + props.content.slice(rightPart[1], end),
				};
			}

			const leftPartMiddle = Math.floor((leftPart[0] + leftPart[1]) / 2);
			const rightPartMiddle = Math.ceil((rightPart[0] + rightPart[1]) / 2);

			calculateText.value = props.content.slice(0, leftPartMiddle) + dot + actionText + dot + props.content.slice(rightPartMiddle, end);

			await waitingScreenRedrawn();

			if (calculate.value!.offsetHeight <= maxHeight) {
				return cuttingMiddleText([leftPartMiddle, leftPart[1]], [rightPart[0], rightPartMiddle]);
			} else {
				return cuttingMiddleText([leftPart[0], leftPartMiddle], [rightPartMiddle, rightPart[1]]);
			}
		};

		const cuttingText: (left: number, right: number) => Promise<{ startText?: string; endText?: string }> = async (left: number, right: number) => {
			const actionText = props.showExpand ? (state.expanded ? props.collapseText : props.expandText) : '';
			const end = props.content.length;

			if (right - left <= 1) {
				if (props.direction === 'end') {
					return { endText: props.content.slice(0, left) + dot };
				} else {
					return { startText: dot + props.content.slice(right, end) };
				}
			}

			const middle = Math.round((left + right) / 2);
			if (props.direction === 'end') {
				calculateText.value = props.content.slice(0, middle) + dot + actionText;
			} else {
				calculateText.value = actionText + dot + props.content.slice(middle, end);
			}

			await waitingScreenRedrawn();

			if (calculate.value!.offsetHeight <= maxHeight) {
				if (props.direction === 'end') {
					return cuttingText(middle, right);
				} else {
					return cuttingText(left, middle);
				}
			} else {
				if (props.direction === 'end') {
					return cuttingText(left, middle);
				} else {
					return cuttingText(middle, right);
				}
			}
		};

		const handlerClick = (type: number) => {
			if (type === 1) {
				state.expanded = true;
			} else if (type === 2) {
				state.expanded = false;
			}
		};

		watch(
			() => [props.rows, props.direction, props.content],
			(newVal, oldVal) => {
				if (newVal != oldVal) calcEllipse();
			}
		);

		onMounted(() => calcEllipse());

		return { bem, ellipsis, calculate, calculateText, isCalculate, ...toRefs(state), handlerClick };
	},
});
</script>
