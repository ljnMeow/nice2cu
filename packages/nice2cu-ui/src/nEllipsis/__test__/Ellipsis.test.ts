import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import { createApp, nextTick } from 'vue';

import nEllipsis from '../';
import nEllipsisVue from '../Ellipsis.vue';

const content =
	'先帝创业未半而中道崩殂，今天下三分，益州疲弊，此诚危急存亡之秋也。然侍卫之臣不懈于内，忠志之士忘身于外者，盖追先帝之殊遇，欲报之于陛下也。诚宜开张圣听，以光先帝遗德，恢弘志士之气，不宜妄自菲薄，引喻失义，以塞忠谏之路也';

test('test Ellipsis plugin', () => {
	const app = createApp({}).use(nEllipsis);
	expect(app.component((nEllipsis as { name: string }).name)).toBeTruthy();
});

describe('test Ellipsis props', () => {
	test('test Ellipsis rows', async () => {
		const wrapper = mount(nEllipsisVue, {
			props: {
				content: content,
				rows: 3,
			},
		});
		await nextTick();
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
	test('test Ellipsis direction start', async () => {
		const wrapper = mount(nEllipsisVue, {
			props: {
				content: content,
				rows: 3,
				direction: 'start',
			},
		});
		await nextTick();
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
	test('test Ellipsis direction center', async () => {
		const wrapper = mount(nEllipsisVue, {
			props: {
				content: content,
				rows: 3,
				direction: 'center',
			},
		});
		await nextTick();
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
	test('test Ellipsis showExpand', async () => {
		const wrapper = mount(nEllipsisVue, {
			props: {
				content: content,
				rows: 3,
				showExpand: true,
			},
		});
		await nextTick();
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
	test('test Ellipsis expandText && collapseText', async () => {
		const wrapper = mount(nEllipsisVue, {
			props: {
				content: content,
				rows: 3,
				showExpand: true,
				expandText: '打开',
				collapseText: '关闭',
			},
		});
		await nextTick();
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
});
