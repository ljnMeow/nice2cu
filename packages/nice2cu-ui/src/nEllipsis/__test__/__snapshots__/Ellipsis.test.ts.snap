// Vitest Snapshot v1

exports[`test Ellipsis props > test Ellipsis direction center 1`] = `
"<div class=\\"n-ellipsis\\">
  <div class=\\"n-ellipsis--calculate\\">先帝创业未半而中道崩殂，今天下三分，益州疲弊，此诚危急......，恢弘志士之气，不宜妄自菲薄，引喻失义，以塞忠谏之路也</div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`test Ellipsis props > test Ellipsis direction start 1`] = `
"<div class=\\"n-ellipsis\\">
  <div class=\\"n-ellipsis--calculate\\">...先帝之殊遇，欲报之于陛下也。诚宜开张圣听，以光先帝遗德，恢弘志士之气，不宜妄自菲薄，引喻失义，以塞忠谏之路也</div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`test Ellipsis props > test Ellipsis expandText && collapseText 1`] = `
"<div class=\\"n-ellipsis\\">
  <div class=\\"n-ellipsis--calculate\\">先帝创业未半而中道崩殂，今天下三分，益州疲弊，此诚危急存亡之秋也。然侍卫之臣不懈于内，忠志之士忘身于外者，盖追...打开</div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`test Ellipsis props > test Ellipsis rows 1`] = `
"<div class=\\"n-ellipsis\\">
  <div class=\\"n-ellipsis--calculate\\">先帝创业未半而中道崩殂，今天下三分，益州疲弊，此诚危急存亡之秋也。然侍卫之臣不懈于内，忠志之士忘身于外者，盖追...</div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`test Ellipsis props > test Ellipsis showExpand 1`] = `
"<div class=\\"n-ellipsis\\">
  <div class=\\"n-ellipsis--calculate\\">先帝创业未半而中道崩殂，今天下三分，益州疲弊，此诚危急存亡之秋也。然侍卫之臣不懈于内，忠志之士忘身于外者，盖追...展开</div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;
